unison-cli/src/Unison/CommandLine/InputPatterns.hs

```diff
@@ -2306,163 +2301,75 @@ commandNameArg =
       globTargets = mempty
     }
 
-exactDefinitionOrPathArg :: ArgumentType
-exactDefinitionOrPathArg =
+exactDefinitionArg :: ArgumentType
+exactDefinitionArg =
   ArgumentType
-    { typeName = "definition or path",
-      suggestions =
-        allCompletors
-          [ termCompletor exactComplete,
-            typeCompletor exactComplete,
-            pathCompletor exactComplete (Set.map Path.toText . Branch.deepPaths)
-          ],
+    { typeName = "definition",
+      suggestions = prefixCompleteTermOrType,
       globTargets = Set.fromList [Globbing.Term, Globbing.Type, Globbing.Namespace]
     }
 
--- todo: improve this
 fuzzyDefinitionQueryArg :: ArgumentType
 fuzzyDefinitionQueryArg =
   ArgumentType
     { typeName = "fuzzy definition query",
-      suggestions =
-        bothCompletors
-          (termCompletor fuzzyComplete)
-          (typeCompletor fuzzyComplete),
+      suggestions = prefixCompleteTermOrType,
       globTargets = Set.fromList [Globbing.Term, Globbing.Type]
     }
 
 definitionQueryArg :: ArgumentType
-definitionQueryArg = fuzzyDefinitionQueryArg {typeName = "definition query"}
+definitionQueryArg = exactDefinitionArg {typeName = "definition query"}
 
 exactDefinitionTypeQueryArg :: ArgumentType
 exactDefinitionTypeQueryArg =
   ArgumentType
     { typeName = "type definition query",
-      suggestions = typeCompletor exactComplete,
+      suggestions = prefixCompleteType,
       globTargets = Set.fromList [Globbing.Type]
     }
 
 exactDefinitionTermQueryArg :: ArgumentType
 exactDefinitionTermQueryArg =
   ArgumentType
     { typeName = "term definition query",
-      suggestions = termCompletor exactComplete,
+      suggestions = prefixCompleteTerm,
       globTargets = Set.fromList [Globbing.Term]
     }
 
-typeCompletor ::
-  Applicative m =>
-  (String -> [String] -> [Completion]) ->
-  String ->
-  Codebase m v a ->
-  Branch.Branch m ->
-  Path.Absolute ->
-  m [Completion]
-typeCompletor filterQuery = pathCompletor filterQuery go
-  where
-    go = Set.map HQ.toText . R.dom . Names.hashQualifyTypesRelation . Names.types . Branch.toNames
-
-termCompletor ::
-  Applicative m =>
-  (String -> [String] -> [Completion]) ->
-  String ->
-  Codebase m v a ->
-  Branch.Branch m ->
-  Path.Absolute ->
-  m [Completion]
-termCompletor filterQuery = pathCompletor filterQuery go
-  where
-    go = Set.map HQ.toText . R.dom . Names.hashQualifyTermsRelation . Names.terms . Branch.toNames
-
 patchArg :: ArgumentType
 patchArg =
   ArgumentType
     { typeName = "patch",
-      suggestions =
-        pathCompletor
-          exactComplete
-          (Set.map Name.toText . Map.keysSet . Branch.deepEdits),
+      suggestions = prefixCompletePatch,
       globTargets = Set.fromList []
     }
 
-allCompletors ::
```

Comment:
Deleted, it's more efficient to produce all the types of completions in a single traversal instead.